src/core/qgsmaplayer.h

```diff
@@ -73,13 +73,17 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     QgsMapLayer( QgsMapLayer::LayerType type = VectorLayer, const QString &name = QString::null, const QString &source = QString::null );
 
-    virtual ~QgsMapLayer();
+    /** Copy constructor.
+     * @param mapLayer the map layer to copy
+     * @param deep to copy unique id if true
+     */
+    QgsMapLayer( const QgsMapLayer &mapLayer, bool deep = true );
```

Comment:
I'm not in favour of this approach. I'd rather see a QgsVectorLayer/QgsRasterLayer ::clone() method and leave the copy constructor /assignment operator deleted.

Reasons include:
- matches the standard set in other parts of the QGIS code (eg symbology)
- avoid unintentional creation of copies
- makes it clear that the result is a new QObject, and will need to be reconnected to any relevant signals/etc